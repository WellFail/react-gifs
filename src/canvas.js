import React, { useEffect, useLayoutEffect, useRef, forwardRef } from "react";
import { createSingleton } from "./hooks";

const calcArgs = (fit, frameSize, canvasSize) => {
  switch (fit) {
    case "fill":
      return [
        0,
        0,
        frameSize.width,
        frameSize.height,
        0,
        0,
        canvasSize.width,
        canvasSize.height,
      ];

    case "contain": {
      const ratio = Math.min(
        canvasSize.width / frameSize.width,
        canvasSize.height / frameSize.height
      );
      const centerX = (canvasSize.width - frameSize.width * ratio) / 2;
      const centerY = (canvasSize.height - frameSize.height * ratio) / 2;
      return [
        0,
        0,
        frameSize.width,
        frameSize.height,
        centerX,
        centerY,
        frameSize.width * ratio,
        frameSize.height * ratio,
      ];
    }

    case "cover": {
      const ratio = Math.max(
        canvasSize.width / frameSize.width,
        canvasSize.height / frameSize.height
      );
      const centerX = (canvasSize.width - frameSize.width * ratio) / 2;
      const centerY = (canvasSize.height - frameSize.height * ratio) / 2;
      return [
        0,
        0,
        frameSize.width,
        frameSize.height,
        centerX,
        centerY,
        frameSize.width * ratio,
        frameSize.height * ratio,
      ];
    }

    default:
      return [0, 0];
  }
};

const combine = (...refs) => (value) => {
  refs.forEach((ref) => {
    if (typeof ref === "function") {
      ref(value);
    } else if (ref != null) {
      ref.current = value;
    }
  });
};

const useCanvasSingleton = createSingleton(() => {
  const canvas = document.createElement("canvas");
  const ctx = canvas.getContext("2d");

  canvas.width = 0;
  canvas.height = 0;

  return ctx;
});

export const Canvas = forwardRef(function Canvas(
  { index, frames, width, height, fit, className, style },
  ref
) {
  const canvasRef = useRef();
  const ctx = useRef();
  const tempCtx = useCanvasSingleton();

  useLayoutEffect(() => {
    if (canvasRef.current) {
      ctx.current = canvasRef.current.getContext("2d");
    }
  }, [canvasRef]);

  useLayoutEffect(() => {
    if (canvasRef.current) {
      canvasRef.current.width = width;
      canvasRef.current.height = height;
    }
  }, [canvasRef, width, height]);

  useEffect(() => {
    const imageData = frames[index];
    if (imageData) {
      if (
        tempCtx.canvas.width < imageData.width ||
        tempCtx.canvas.height < imageData.height
      ) {
        tempCtx.canvas.width = imageData.width;
        tempCtx.canvas.height = imageData.height;
      }
      if (width > 0 && height > 0) {
        ctx.current.clearRect(0, 0, width, height);
        tempCtx.clearRect(0, 0, tempCtx.canvas.width, tempCtx.canvas.height);
      }

      tempCtx.putImageData(imageData, 0, 0);
      ctx.current.drawImage(
        tempCtx.canvas,
        ...calcArgs(fit, imageData, { width, height })
      );
    }
  }, [index, frames, width, height, fit]);

  return (
    <canvas ref={combine(canvasRef, ref)} className={className} style={style} />
  );
});
