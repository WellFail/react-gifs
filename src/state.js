import { useReducer } from "react";

const reducer = (state, action = {}) => {
  if (typeof action === "function") {
    action = action(state);
  }

  const isLoadedChange = action.loaded && state.loaded !== action.loaded
  const { playing, ...rest } = action;
  const copy = { ...state, ...rest };

  if (isLoadedChange) {
    Object.assign(copy, {
      playing: copy.autoPlay && copy.loaded,
    });
  }
  
  if (action.delays != null || action.frames != null) {
    Object.assign(copy, {
      length: copy.frames.length,
    });

    if (process.env.NODE_ENV !== "production") {
      if (copy.frames.length !== copy.delays.length)
        throw Error("frames and delays have different sizes");
    }
  }

  if (action.index != null || action.frames != null) {
    Object.assign(copy, {
      index: copy.length === 0 ? 0 : (copy.length + copy.index) % copy.length,
    });
  }

  if (action.playing != null) {
    Object.assign(copy, copy.loaded ? { playing } : { autoPlay: playing });
  }

  return copy;
};

const initializer = (stateOrFn) =>
  reducer(
    {
      autoPlay: true,
      playing: false,
      frames: [],
      delays: [],
      index: 0,
      length: 0,
      loaded: false,
    },
    stateOrFn
  );

const usePlayerState = (stateOrFn) => {
  return useReducer(reducer, stateOrFn, initializer);
};

export { reducer, initializer, usePlayerState };
