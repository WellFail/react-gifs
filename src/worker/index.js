import code from './source'

export default function WorkerFactory (options) {
  const blob = new Blob([code], { type: 'application/javascript' });
  const url = URL.createObjectURL(blob);
  const worker = new Worker(url, options)
  URL.revokeObjectURL(url)

  return worker 
}
