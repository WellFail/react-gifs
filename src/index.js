import { usePlayerState } from "./state";
import { Canvas } from "./canvas";
import { useWorkerParser, useParser, usePlayback } from "./hooks";

export { useWorkerParser, useParser, Canvas, usePlayback, usePlayerState };
