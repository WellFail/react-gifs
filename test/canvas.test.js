import "jest-canvas-mock";
import React, { createRef } from "react";
import { render } from "@testing-library/react";
import { Canvas } from "../src/canvas";

const createImageData = (width = 10, height = 10) => {
  let data = new Uint8ClampedArray(width * height * 4);
  for (let i = 3; i < data.length; i += 4) {
    data[i] = 255;
  }
  return new ImageData(data, width, height);
};

const getDraws = (containerORref) => {
  const canvas =
    "current" in containerORref
      ? containerORref.current
      : containerORref.querySelector("canvas");
  const ctx = canvas.getContext("2d");

  return ctx.__getDrawCalls();
};

it("works as expected", () => {
  const ref = createRef();
  const frames = [createImageData(10, 10), createImageData(20, 20)];

  const { rerender } = render(
    <Canvas
      index={0}
      frames={frames}
      ref={ref}
      width={20}
      height={20}
      fit="fill"
    />
  );
  rerender(
    <Canvas
      index={1}
      frames={frames}
      ref={ref}
      width={20}
      height={20}
      fit="fill"
    />
  );

  const calls = getDraws(ref);
  expect(calls).toMatchSnapshot();
});

it("works without ref", async () => {
  const frames = [createImageData(10, 10)];
  const { container } = render(<Canvas index={0} frames={frames} />);

  const draws = getDraws(container);
  expect(draws.length).toEqual(1);
});

it("works with callback ref", async () => {
  const refCallback = jest.fn();
  const frames = [];
  render(<Canvas index={0} frames={frames} ref={refCallback} />);

  expect(refCallback).toHaveBeenCalled();
  expect(refCallback.mock.calls[0][0]).toBeInstanceOf(HTMLCanvasElement);
});
