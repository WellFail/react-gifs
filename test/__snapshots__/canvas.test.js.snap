// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works as expected 1`] = `
Array [
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 20,
      "dWidth": 20,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="20"
        width="20"
      />,
      "sHeight": 10,
      "sWidth": 10,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 20,
      "dWidth": 20,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="20"
        width="20"
      />,
      "sHeight": 20,
      "sWidth": 20,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
